### PR TITLE
unsafe_merge: detect value with no root

### LINKED ISSUE #1087: `unsafe_merge` crashes with nested structured config and union type
**Describe the bug**
`unsafe_merge` crashes with nested structured config and union type.

**To Reproduce**
```python
from dataclasses import dataclass

from omegaconf import OmegaConf

@dataclass
class A:
    x: int | str

@dataclass
class B:
    a: A

def main():
    x = OmegaConf.unsafe_merge(OmegaConf.structured(B), {'a': {'x': 1}})
    print(x)

if __name__ == '__main__':
    main()
```

**Expected behavior**
`{'a': {'x': 1}}` will be printed, just like `OmegaConf.merge`.

**Actual result**
```
Traceback (most recent call last):
  File "./test.py", line 18, in <module>
    main()
  File "./test.py", line 14, in main
    x = OmegaConf.unsafe_merge(OmegaConf.structured(B), {'a': {'x': 1}})
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../omegaconf/omegaconf.py", line 308, in unsafe_merge
    target.merge_with(*configs[1:])
  File ".../omegaconf/basecontainer.py", line 492, in merge_with
    self._format_and_raise(key=None, value=None, cause=e)
  File ".../omegaconf/base.py", line 231, in _format_and_raise
    format_and_raise(
  File ".../omegaconf/basecontainer.py", line 490, in merge_with
    self._merge_with(*others)
  File ".../omegaconf/basecontainer.py", line 514, in _merge_with
    BaseContainer._map_merge(self, other)
  File ".../omegaconf/basecontainer.py", line 399, in _map_merge
    dest_node._merge_with(src_node)
  File ".../omegaconf/basecontainer.py", line 514, in _merge_with
    BaseContainer._map_merge(self, other)
  File ".../omegaconf/basecontainer.py", line 358, in _map_merge
    expand(dest)
  File ".../omegaconf/basecontainer.py", line 342, in expand
    node._set_value(val)
  File ".../omegaconf/dictconfig.py", line 647, in _set_value
    raise e
  File ".../omegaconf/dictconfig.py", line 644, in _set_value
    self._set_value_impl(value, flags)
  File ".../omegaconf/dictconfig.py", line 677, in _set_value_impl
    self.__setitem__(k, v)
  File ".../omegaconf/dictconfig.py", line 314, in __setitem__
    self._format_and_raise(key=key, value=value, cause=e)
  File ".../omegaconf/base.py", line 231, in _format_and_raise
    format_and_raise(
  File ".../omegaconf/dictconfig.py", line 308, in __setitem__
    self.__set_impl(key=key, value=value)
  File ".../omegaconf/dictconfig.py", line 318, in __set_impl
    self._set_item_impl(key, value)
  File ".../omegaconf/basecontainer.py", line 535, in _set_item_impl
    if self._get_root() is value._get_root():
                           ^^^^^^^^^^^^^^^^^
  File ".../omegaconf/base.py", line 289, in _get_root
    assert isinstance(self, Container)
AssertionError
```

**Additional context**
- [x] OmegaConf version: 2.3.0
- [x] Python version: 3.11.4
- [x] Operating system: Ubuntu 22.04

### PR DESCRIPTION
**Pull request by Jasha10**
Closes #1087.

This PR addresses an edge-case where `AssertionError` is raised during a
call to `OmegaConf.unsafe_merge`.

This PR is marked as a draft:
- I haven't written tests yet
- I'm not sure if this is the best approach to solving the issue.

### CODE REVIEW

File: omegaconf/basecontainer.py [hunk 1]
```diff
@@ -531,8 +531,12 @@ def _set_item_impl(self, key: Any, value: Any) -> None:
         if isinstance(value, Node):
             do_deepcopy = not self._get_flag("no_deepcopy_set_nodes")
             if not do_deepcopy and isinstance(value, Box):
+                # Detect the special case where value has no root:
+                value_has_root = (
+                    isinstance(value, Container) or value._get_parent() is not None
+                )
                 # if value is from the same config, perform a deepcopy no matter what.
-                if self._get_root() is value._get_root():
+                if value_has_root and self._get_root() is value._get_root():
```
Comment (Jasha10) [inline omegaconf/basecontainer.py:539]: Normally, every node should have a root. The "root" of a node is a top-level `Container` that owns the node. In other words, the root is the top of the config tree.

In example from issue #1087, the call here to `value._get_root()` causes an `AssertionError`. This happens because `value` is a temporary `UnionNode` value that has no parent (and thus has no root).

Comment (omry) [reply omegaconf/basecontainer.py:539]: Both approaches seems okay to me.
Also, please add a test for unsafe_merge similar in this case (similar to the repro in the bug report).

Comment (omry) [reply omegaconf/basecontainer.py:539]: Make sure that your proposed logic does not treat the nodes as being from the same config (see comment in 538).

Comment (Jasha10) [reply omegaconf/basecontainer.py:539]: I've added the test.

> Make sure that your proposed logic does not treat the nodes as being from the same config 

Yup. If `value_has_root` is `False` then `value` is a temporary node and does not come from the same config as `self`.

**Comment by mohammad7t** [comment]
Hi @omry ,

Thank you for creating this PR. I'm wondering if there is any blocker for merging this PR, that I could potentially help resolving?

**Comment by omry** [comment]
Hi @Jasha10.
Thanks for this. I think I have a cleaner solution in #1285.
